### ISSUE TITLE
Add new group members to Google Group [2]

### ISSUE THREAD

**Issue by Matt-Holland** [affinityworks/main#532]
# Acceptance Criteria

GIVEN I am signing up for a group with gsuite enabled:
* WHEN I click "submit" on the subgroup signup form
* THEN I will be added to the group's google-group

# Context

Up until now, we have shown a "Google Groups -- coming soon" teaser in the links section of a Group homepage. With this card, we can show the actual link and play #578

**Comment by vaughanj10** [comment]
Questions for @aguestuser:
- Like the previous Google Groups card for adding the owner of a subgroup to a google group, this card calls for any member signup to a subgroup to be added to the google group. However, this requires going through the Signups controller, which as far as I can tell is generalized to be for any group or subgroup. Do we have a mechanism currently to determine if a group is a subgroup?

**Comment by Matt-Holland** [comment]
This probably bears some deeper discussion, but:
- Every Group in Affinity should be essentially the same in terms of it's qualities and abilities, _unless_ they have been specifically de-activated for some reason.
- To wit: a "subgroup" is sort of a hack on the original model in the first place, since the fundamental design of Affinity is a web of relationships not a hierarchy. A hierarchy, and therefore a "subgroup" is just a special case of the general design. So for instance as currently designed, a subgroup *could* create it's _own_ sub-subgroups, ad infinitum. Our concept of a "network" could be thought of (I think??) as a set of rules for creating a particular set of relationships among groups. I.e., there's nothing magical about the top group in a "network" except that all of its subgroups publish themselves to it, and the top group does not publish itself back to anyone.
- Therefore, I _think_ it's true that anything any group can do (e.g. trigger automatic Google Group creation), _every_ group should be able to do (unless it's explicitly toggled 'off').

@aguestuser , I'm especially interested in your thoughts about if the above makes sense...

**Comment by aguestuser** [comment]
Geeze, lots of interesting stuff here. Thoughts:

Straight-forward things:

* @vaughanj10 you are correct in your reading that the signups controller can be called for any group, as a signup is a resource that joins a group to a person to a signup form (and as the `signups` route is declared in `routes.rb` as a subresource of `group`)
* i think the logic that @Matt-Holland outlines is correct and that for this cards, "subgroups" is just a special case of "groups" and that the title of this card should actually be "when a new member signs up for a group, add them to the group's google group"

Less straightforward clarification:

>I think it's true that anything any group can do (e.g. trigger automatic Google Group creation), every group should be able to do (unless it's explicitly toggled 'off').

This is not currently the case with regards Google Groups and I want to make sure we're on the same page. The way you describe capabilities makes it sound like we only have `opt-out` feature toggles. In the case of Google Groups, we also have `opt-in` feature toggles. This is just to say that no group in Affinity (as currently implemented) will ever have the Google Group functionality turned on unless they explicitly ask for it.

The reasoning is that in order to activiate a google group there is a 37-step manual process that a client has to go through in order to generate the google group and the credentails json that we need to interact with the API. Since for most groups we don't have that, for most groups the feature flag is set to off -- so that (among other things) we won't try to perform API calls that won't work and/or display UI elements that won't make any sense (since they won't have the google group that the UI would be talking about.)

If this is bad and we don't want it -- ie: if we want it to be actually the case that all groups have the same capabilities unless a feature is turned *off* -- then, as i see it, we could do 1 of 2 things:

1. sprinkle a lot of defensive code throughout the codebase to make it degrade gracefully in the case where it can't find the API credentials json file (and thus sort of insert a bunch of implicit feature flags into the code: "if you can't find this JSON file, the feature is "off")
2. figure out a way to automate those 37 steps (perhaps via a web scraping script to automate clicking all the buttons, etc., if there is no API for them) so that we can actually automatically create a google group for any group that signs up and thus eliminate the need for feature flags (NOTE: this would probably require storing the JSON in our database, since, as I just discovered Friday, we may not write files to our app's filesystem on heroku)

Neither option is completely simple. I'm curious to hear what other folks think about the trade-offs.

For the purposes of this card, I don't think we need to make a decision to proceed.

**Comment by aguestuser** [comment]
resolved by #535

**Comment by aguestuser** [comment]
whoops. not closed!

**Comment by Matt-Holland** [comment]
Doesn't seem to work on master.

```2018-04-12T06:23:13.413384+00:00 app[web.1]: D, [2018-04-12T06:23:13.413302 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   [1m[36mNetworkMembership Load (1.1ms)[0m  [1m[34mSELECT  "network_memberships".* FROM "network_memberships" WHERE "network_memberships"."group_id" = $1 AND "network_memberships"."primary" = $2 ORDER BY "network_memberships"."id" ASC LIMIT $3[0m  [["group_id", 24], ["primary", "t"], ["LIMIT", 1]]
2018-04-12T06:23:13.415523+00:00 app[web.1]: D, [2018-04-12T06:23:13.415429 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   [1m[36mNetwork Load (0.9ms)[0m  [1m[34mSELECT  "networks".* FROM "networks" WHERE "networks"."id" = $1 LIMIT $2[0m  [["id", 2], ["LIMIT", 1]]
2018-04-12T06:23:13.416718+00:00 app[web.1]: D, [2018-04-12T06:23:13.416631 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   [1m[36mCACHE (0.0ms)[0m  [1m[34mSELECT  "network_memberships".* FROM "network_memberships" WHERE "network_memberships"."group_id" = $1 AND "network_memberships"."primary" = $2 ORDER BY "network_memberships"."id" ASC LIMIT $3[0m  [["group_id", 24], ["primary", "t"], ["LIMIT", 1]]
2018-04-12T06:23:13.417763+00:00 app[web.1]: D, [2018-04-12T06:23:13.417680 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   [1m[36mCACHE (0.0ms)[0m  [1m[34mSELECT  "networks".* FROM "networks" WHERE "networks"."id" = $1 LIMIT $2[0m  [["id", 2], ["LIMIT", 1]]
2018-04-12T06:23:13.420698+00:00 app[web.1]: D, [2018-04-12T06:23:13.420601 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   [1m[36mGoogleGroup Load (1.0ms)[0m  [1m[34mSELECT  "google_groups".* FROM "google_groups" WHERE "google_groups"."group_id" = $1 LIMIT $2[0m  [["group_id", 24], ["LIMIT", 1]]
2018-04-12T06:23:13.421824+00:00 app[web.1]: D, [2018-04-12T06:23:13.421746 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] Sending HTTP get https://www.googleapis.com/admin/directory/v1/groups/?
2018-04-12T06:23:13.574046+00:00 app[web.1]: D, [2018-04-12T06:23:13.573892 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] 400
2018-04-12T06:23:13.574183+00:00 app[web.1]: D, [2018-04-12T06:23:13.574103 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] #<Hurley::Response GET https://www.googleapis.com/admin/directory/v1/groups/ == 400 (171 bytes) 41ms>
2018-04-12T06:23:13.574492+00:00 app[web.1]: D, [2018-04-12T06:23:13.574413 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] Caught error badRequest: Bad Request
2018-04-12T06:23:13.574688+00:00 app[web.1]: D, [2018-04-12T06:23:13.574606 #4] DEBUG -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] Error - #<Google::Apis::ClientError: badRequest: Bad Request>
2018-04-12T06:23:13.574698+00:00 app[web.1]: 
2018-04-12T06:23:13.575507+00:00 app[web.1]: I, [2018-04-12T06:23:13.575423 #4]  INFO -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] Completed 500 Internal Server Error in 248ms (ActiveRecord: 33.5ms)
2018-04-12T06:23:13.578196+00:00 app[web.1]: F, [2018-04-12T06:23:13.578092 #4] FATAL -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   
2018-04-12T06:23:13.578282+00:00 app[web.1]: F, [2018-04-12T06:23:13.578209 #4] FATAL -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] Google::Apis::ClientError (badRequest: Bad Request):
2018-04-12T06:23:13.578365+00:00 app[web.1]: F, [2018-04-12T06:23:13.578294 #4] FATAL -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6]   
2018-04-12T06:23:13.578461+00:00 app[web.1]: F, [2018-04-12T06:23:13.578387 #4] FATAL -- : [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] app/services/google_api/fetch_google_group.rb:6:in `call'
2018-04-12T06:23:13.578463+00:00 app[web.1]: [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] app/services/google_api/service.rb:47:in `fetch_google_group'
2018-04-12T06:23:13.578465+00:00 app[web.1]: [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] app/services/signups/after_create.rb:9:in `call'
2018-04-12T06:23:13.578466+00:00 app[web.1]: [2793de35-8d9c-4f96-a93f-63d8efa9e3e6] app/controllers/signups_controller.rb:17:in `create'```

**Comment by aguestuser** [comment]
hey @Matt-Holland on initial attempts, i am unable to reproduce this bug on `dev-affinityworks`. if i try the following it works:

* create a fresh group called `My Fun Group` with `austin@affinity.works` as the email of the organizer
  * -> i see a google group gets created and that `austin@affinity.works` is its owner
* join the group as `guest.austin@gmail.com`
  * -> i see that `guest.austin@gmail.com` was added to the google group

if, however, i try to join as `guest.austin+1@gmail.com` i get a halting error that is a *different* error than the one you show above, but which is also one we cannot avoid, since the error is due to Google Groups not being able to find the email address `guest.austin+1@gmail.com`:

```
2018-04-12T07:31:13.898668+00:00 app[web.1]: D, [2018-04-12T07:31:13.898539 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] 200
2018-04-12T07:31:13.898751+00:00 app[web.1]: D, [2018-04-12T07:31:13.898688 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] #<Hurley::Response GET https://www.googleapis.com/admin/directory/v1/groups/034g0dwd0rs5de6 == 200 (402 bytes) 104ms>
2018-04-12T07:31:13.899868+00:00 app[web.1]: D, [2018-04-12T07:31:13.899802 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Success - #<Google::Apis::AdminDirectoryV1::Group:0x00000007770838
2018-04-12T07:31:13.899870+00:00 app[web.1]:  @admin_created=true,
2018-04-12T07:31:13.899872+00:00 app[web.1]:  @description="This is a Google Group generated by Affinity.works.",
2018-04-12T07:31:13.899873+00:00 app[web.1]:  @direct_members_count="2",
2018-04-12T07:31:13.899875+00:00 app[web.1]:  @email="my-fun-group@affinity.works",
2018-04-12T07:31:13.899876+00:00 app[web.1]:  @etag="\"HIWxtAjmBmqPQjjDV2Duo181uSc/SiOGOEiSNxX5o_UmzX2YhUuV65U\"",
2018-04-12T07:31:13.899877+00:00 app[web.1]:  @id="034g0dwd0rs5de6",
2018-04-12T07:31:13.899878+00:00 app[web.1]:  @kind="admin#directory#group",
2018-04-12T07:31:13.899880+00:00 app[web.1]:  @name="My Fun Group",
2018-04-12T07:31:13.899881+00:00 app[web.1]:  @non_editable_aliases=["my-fun-group@affinity.works.test-google-a.com"]>
2018-04-12T07:31:13.899882+00:00 app[web.1]:
2018-04-12T07:31:13.900804+00:00 app[web.1]: D, [2018-04-12T07:31:13.900735 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Sending HTTP post https://www.googleapis.com/admin/directory/v1/groups/034g0dwd0rs5de6/members?
2018-04-12T07:31:14.038491+00:00 app[web.1]: D, [2018-04-12T07:31:14.038367 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] 404
2018-04-12T07:31:14.038580+00:00 app[web.1]: D, [2018-04-12T07:31:14.038527 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] #<Hurley::Response POST https://www.googleapis.com/admin/directory/v1/groups/034g0dwd0rs5de6/members == 404 (239 bytes) 133ms>
2018-04-12T07:31:14.038763+00:00 app[web.1]: D, [2018-04-12T07:31:14.038711 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Caught error notFound: Resource Not Found: guest.austin+222@gmail.com
2018-04-12T07:31:14.038893+00:00 app[web.1]: D, [2018-04-12T07:31:14.038840 #4] DEBUG -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Error - #<Google::Apis::ClientError: notFound: Resource Not Found: guest.austin+222@gmail.com>
2018-04-12T07:31:14.038895+00:00 app[web.1]:
2018-04-12T07:31:14.039377+00:00 app[web.1]: I, [2018-04-12T07:31:14.039325 #4]  INFO -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Completed 500 Internal Server Error in 467ms (ActiveRecord: 22.0ms)
2018-04-12T07:31:14.040517+00:00 app[web.1]: F, [2018-04-12T07:31:14.040460 #4] FATAL -- : [f46f90ca-256b-4741-82b1-b99d210b0d09]
2018-04-12T07:31:14.040580+00:00 app[web.1]: F, [2018-04-12T07:31:14.040520 #4] FATAL -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] Google::Apis::ClientError (notFound: Resource Not Found: guest.austin+222@gmail.com):
2018-04-12T07:31:14.040640+00:00 app[web.1]: F, [2018-04-12T07:31:14.040589 #4] FATAL -- : [f46f90ca-256b-4741-82b1-b99d210b0d09]
2018-04-12T07:31:14.040692+00:00 app[web.1]: F, [2018-04-12T07:31:14.040651 #4] FATAL -- : [f46f90ca-256b-4741-82b1-b99d210b0d09] app/services/google_api/add_member_to_google_group.rb:12:in `call'
2018-04-12T07:31:14.040694+00:00 app[web.1]: [f46f90ca-256b-4741-82b1-b99d210b0d09] app/services/google_api/service.rb:57:in `add_member_to_google_group'
2018-04-12T07:31:14.040695+00:00 app[web.1]: [f46f90ca-256b-4741-82b1-b99d210b0d09] app/services/signups/after_create.rb:10:in `call'
2018-04-12T07:31:14.040696+00:00 app[web.1]: [f46f90ca-256b-4741-82b1-b99d210b0d09] app/controllers/signups_controller.rb:17:in `create'
```

notably, if i try to join as `austin+222@affinity.works` it succeeds and i am added to the group.

so the bug seems to be a *testing* bug insofar as google groups is unable to add vanilla gmail addresses with suffixes of the form `foo+1@gmail.com` whereas it *is* able to add google apps email addresses wth suffices of the form `foo+1@somedomain.com`. 

can you verify whether you were trying with `matt.holland+N@gmail.com` and whether this might be the cause of the error? can you try instead with an `affinity.works` domain and see if this fixes it?

otherwise, i cannot explain the error you are seeing, and cannot reproduce it on heroku dev.

**Comment by aguestuser** [comment]
aha!

i *can* reproduce the exact error you found if i try to sign up for a group that (1) has google groups toggled on, but (2) was created before we introduced the logic that stores google group ids so we can use the id to fetch the google group from the API in the post-signup hook.

to fix this we could either: (1) run a migration to fetch and persist the google group ids for all pre-existing groups-with-google groups that we have created in the last 2 weeks before this PR, (2) provide a nil-check that will prevent the program from erroring in such cases but not add new members to the google group, (3) blast away the database on SL prod (the only place where google-group-enabled groups currently exist, but which is not yet in active use) so that the error will never occur there.

(1) is annoying and complicated.

i can push a fix that does (2) right now, then we can do (3) once it's promoted to prod.

sound good?

if i do (2), can you verfiy that you no longer get errors if you try to signup for an old group (but are not added to the google group) but that you *can* both not get errors and get added to a google group if you try with a freshly created group?

**Comment by aguestuser** [comment]
fix pushed. here are next steps as outlined in its commit message:

```
* this addresses bug discovered by @Matt-Holland when trying to signup
  for a group that is in a network with google groups toggled on, but
  which does not have a `GoogleGroup` model (and hence a `group_key`)
  persisted in the DB
* this fix prevents the program from throwing a halting error when
  such a use case happens
* it does *NOT* enable the new signup to be added to the google group
* to do that, we would have to run a migration to retrieve the ids of
  all groups that *have* google groups and create new `GoogleGroup`
  instances for them
* instead of that, i'd rather ship with this fix, and blast away all
  the groups in the beta version of the SL prod app
* if i am wrong about this, we can create the migration. it will probably
  take a few hours to get right
```

**Comment by aguestuser** [comment]
closing as resolved by #615

**Comment by aguestuser** [comment]
heads up: because the PR was merged before approved (i think b/c @Matt-Holland was looking at old groups and not seeing the feature in action and thought merging would help), the push i made to fix the bug discovered above (a86cb9d) never made it into master, and there is still a halting error for old groups on master. i'm going to open a separate PR with that fix, then merge and promote it myself.
